### ISSUE TITLE
some doubt about cosine similarity

### ISSUE THREAD

**Issue by caiqi** [zergylord/oneshot#2]
Hi, it seems that the cosine distance does not take x_hat's magnitude into consideration. The magnitude of x_hat is commented. Does that make difference on final performance? Thanks.

**Comment by zergylord** [comment]
Thanks for the question. Yes, it turns down that normalizing wrt x_hat degrades performance significantly. Indeed, I emailed the author, and he confirmed this 'one sided' cosine similarity is what they used. It makes sense in that since all comparisons are relative to x_hat (and are later softmax'd), you don't need to normalize it. Hope that helps!

**Comment by caiqi** [comment]
Thanks for the response. I spent days trying to improve the performance with normal cosine similarity and found that tuning the temperate of softmax could improve the performance slightly. It seems that softmax expects non normalized logits,  while normal cosine similarity is normalized to [-1,1]. Hope it may be useful to others!
